pep-9999.rst

```diff
@@ -0,0 +1,503 @@
+authors tend to not keep multiple interpreters in mind when developing,
+and it is currently cumbersome to test the behavior.
+
+Rationale for Per-module State
+------------------------------
+
+Instead of focusing on per-interpreter state, Python's C API is evolving
+to better support the more granular *per-module* state. By default,
+C-level data will be attached to a *module object*. Each interpreter
+will then create its own module object, keeping data separate. For
+testing the isolation, multiple module objects corresponding to a single
+extension can even be loaded in a single interpreter.
+
+Per-module state provides an easy way to think about lifetime and
+resource ownership: the extension module author will set up when a
+module object is created, and clean up when it's freed. In this regard,
+a module is just like any other ``PyObject *``; there are no “on
+interpreter shutdown” hooks to think about—or forget about.
+
+Goal: Easy-to-use Module State
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+It is currently cumbersome or impossible to do everything the C API
+offers while keeping modules isolated. Enabled by PEP 384, changes in
+PEPs 489 and 573 (and future planned ones) aim to first make it
+*possible* to build modules this way, and then to make it *easy* to
+write new modules this way and to convert old ones, so that it can
+become a natural default.
+
+Even if per-module state becomes the default, there will be use cases
+for different levels of encapsulation: per-process, per-interpreter,
+per-thread or per-task state. The goal is to treat these as exceptional
+cases: they should be possible, but extension authors will need to
+think more carefully about them.
+
+Non-goals: Speedups and the GIL
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There is some effort to speed up CPython on multi-core CPUs by making the GIL
+per-interpreter. While isolating interpreters helps that effort,
+defaulting to per-module state will be beneficial even if no speedup is
+achieved, as it makes supporting multiple interpreters safer by default.
+
+How to make modules safe with multiple interpreters
+---------------------------------------------------
+
+There are many ways to correctly support multiple interpreters in
+extension modules. The rest of this text describes the preferred way to
+write such a module, or to convert an existing module.
+
+Note that support is a work in progress; the API for some features your
+module needs might not yet be ready.
+
+A full example module is currently available in `a fork on
+GitHub <https://github.com/encukou/cpython/blob/xxlimited-facelift/Modules/xxlimited.c>`__.
+
+.. XXX: Later, it should be in the CPython source tree.
+
+This section assumes that “*you*” are an extension module author.
+
+
+Isolated Module Objects
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The key point to keep in mind when developing an extension module is
+that several module objects can be created from a single shared library.
+For example:
+
+.. code:: pycon
```

Comment:
`pycon` or `python`?

Reply:
I intended to use the `pycon` lexer; short for [Python Console](https://pygments.org/docs/lexers/#pygments.lexers.python.PythonConsoleLexer).